### PR TITLE
Preserve side sets when using MeshGeneratorExtruder

### PR DESCRIPTION
**Pull request by permcody**
This PR fixes the lost sidesets in the MeshExtruderGenerator. It appears that more work needs to be done to preserve node sets in DistributedMesh but that's beyond the scope of this first ticket.

### CODE REVIEW

File: test/tests/meshgenerators/mesh_extruder_generator/tests [hunk 1]
```diff
@@ -50,11 +46,20 @@
...
-      detail = 'where the extrusion occurs based on another generated mesh.'
+      detail = 'where the extrusion occurs based on another generated mesh,'
+    []
+
+    [preserve_sideset_names]
+      type = 'Exodiff'
+      input = 'gen_extrude.i'
+      exodiff = 'gen_extrude_out_left_right.e'
+      cli_args = 'BCs/first/boundary=left BCs/second/boundary=right Outputs/file_base=gen_extrude_out_left_right'
+      mesh_mode = 'REPLICATED' # Node sets still need work in distrubted mesh mode
```
Comment (gambka) [inline test/tests/meshgenerators/mesh_extruder_generator/tests]: distrubted -> distributed

Comment (permcody) [reply test/tests/meshgenerators/mesh_extruder_generator/tests]: Freudian slip

**Comment by lindsayad** [comment]
Does exodiff error out if two files don't have matching sidesets?

**Comment by gambka** [comment]
> Does exodiff error out if two files don't have matching sidesets?

I think it only checks sidesets and nodesets as a part of the exodiff if     

`exodiff_opts = '-pedantic'`

is set in the block for that test in the tests file.

**Comment by permcody** [comment]
Right, it does not error out which is why this subtle bug went unnoticed.
The new test uses “old” side sets explicitly. There is more work to do for
node sets.

**Comment by lindsayad** [comment]
Hmm, I'm not seeing the 'top' sideset in paraview

**Comment by lindsayad** [comment]
false alarm, just had to scroll 😆 but yea it looks like the nodeset names are being lost even in serial with replicated mesh

**Review by lindsayad — APPROVED** [review]

**Comment by permcody** [comment]
Turns out another user reported the issue with missing nodesets/sidesets here: https://github.com/idaholab/moose/issues/13346. I'm not sure if pedantic will cause other issues though. What happens if things get renumbered? I'm willing to give it a shot. Makes me wonder how many time bombs we have lurking in the MeshGenerator system.

**Comment by lindsayad** [comment]
I like how you tested this by making the BCs use the old sideset names. I was going to say that this error is misleading:
```
*** ERROR ***
/Users/lindad/projects/moose/test/tests/meshgenerators/mesh_extruder_generator/gen_extrude.i:41: (BCs/first/boundary):
    the following node set ids do not exist on the mesh: 8
```
because it says node instead of side...***but*** I see that the BC object being used here is actually nodal so the error message is appropriate.

What does seem weird though is that the NodalBC is able to work if it finds a matching ***sideset*** name?? Shouldn't it be looking only at nodesets?

**Comment by lindsayad** [comment]
So I've been doing some testing, and it doesn't appear that `-pedantic` will do name checking/comparisons. For example, these two meshes yielded `exodiff: Files are the same` even with `-pedantic`:

Screen Shot 2019-09-18 at 8 49 06 AM
Screen Shot 2019-09-18 at 8 49 14 AM

**Comment by gambka** [comment]
Yeah, I think `-pedantic` ensures that the IDs are the same.  However, without pedantic it won't even check nodeset or sideset IDs and say they are the same.

**Comment by lindsayad** [comment]
> Yeah, I think `-pedantic` ensures that the IDs are the same. However, without pedantic it won't even check nodeset or sideset IDs and say they are the same.

👍 Nor even in fact if the number of sets is the same!

**Comment by permcody** [comment]
We might need to explore other options for these tests. I wonder if VTK output will do the trick?

**Comment by lindsayad** [comment]
> I wonder if VTK output will do the trick?

🤷‍♂ Sounds like a question for @aeslaughter

**Comment by lindsayad** [comment]
> I believe that's why this is working.

It looks like this works because of [this line](https://github.com/idaholab/moose/blob/next/framework/src/mesh/MooseMesh.C#L1106). `BoundaryInfo::get_id_by_name` queries `_ss_id_to_name` and then `_ns_id_to_name` maps, returning as soon as it finds a match, which in this case is actually from the sideset map. If it queried the nodeset map only, we would be unsuccessful:
```
(lldb) p _ns_id_to_name
(std::__1::map<short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<short>, std::__1::allocator<std::__1::pair<const short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >) $12 = size=2 {
  [0] = (first = 6, second = "new_front")
  [1] = (first = 7, second = "new_back")
}
(lldb) p _ss_id_to_name
(std::__1::map<short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<short>, std::__1::allocator<std::__1::pair<const short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >) $13 = size=6 {
  [0] = (first = 0, second = "bottom")
  [1] = (first = 1, second = "right")
  [2] = (first = 2, second = "top")
  [3] = (first = 3, second = "left")
  [4] = (first = 6, second = "new_front")
  [5] = (first = 7, second = "new_back")
```

**Comment by aeslaughter** [comment]
I have not dealt with VTK output from libMesh directly, so I don't know if it can help.

**Pull request merged**